### PR TITLE
Bn254 precompiles

### PR DESCRIPTION
**Pull request by sisuresh**
### What

An extension of https://github.com/stellar/rs-soroban-env/pull/1597.

### CODE REVIEW

File: soroban-env-host/src/crypto/bn254.rs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+        // Panic analysis:
+        //
+        // The following potential panic conditions could exist:
+        // 1. if two input vector lengths are not equal. There is a `zip_eq`
+        // which panics if the length of the two vectors are not equal. This is
+        // weeded out up front.
+        //
+        // 2. `coeffs.next().unwrap()`. This occurs when the algorithm Loops
+        // over pairs of `(a: G1Affine, b: G2Affine)`, converting them into
+        // `Vec<(G1Prepared, G2Preared::EllCoeff<Config>)>`, the latter contains
```
Comment (dmkozh) [inline soroban-env-host/src/crypto/bn254.rs:319]: typo: G2Prepared

File: soroban-env-host/src/crypto/bn254.rs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        // three elements of Fp2. For each pair, the coeffs.next() can at most
+        // be called twice, when the bit being looped over in `Config::X` is
+        // set. So this panic cannot happen.
+        //
+        // 3. if any of the G1Affine point is infinity. The ell() function which
+        // calls p.xy().unwrap(), which is when the point is infinity. This
+        // condition also cannot happen because when the pairs are generated,
```
Comment (dmkozh) [inline soroban-env-host/src/crypto/bn254.rs:326]: I don't see this actually happening, how exactly do we filter these? I assume `bn254_affine_deserialize` could be returning an error  for the zero points, but that's not filtering. I can also see a test for pairing zero points and it seems like it works, but I don't understand how exactly. 
We also don't seem to have a test where there is a pair of zeros together with non-zero pairs (or at least I couldn't find it).

Comment (sisuresh) [reply soroban-env-host/src/crypto/bn254.rs:326]: Sorry this comment was supposed to be removed. I still need to do panic analysis.

Comment (sisuresh) [reply soroban-env-host/src/crypto/bn254.rs:326]: But I believe the zero filtering happens in the arkworks library.

Comment (dmkozh) [reply soroban-env-host/src/crypto/bn254.rs:326]: That's still a bit suspicious because we reject the empty input vectors, but they might be effectively treated as empty by the arkworks library during pairing and we don't do anything about that. I don't understand the use cases for this well enough, but the inconsistency is a bit concerning. Maybe it's fine though, I don't really understand the context this would be used in, but just wanted to make sure that we're at least intentional about this.

Comment (dmkozh) [reply soroban-env-host/src/crypto/bn254.rs:326]: That's fair, but I still don't fully understand the rationale for prohibiting empty input vectors - you still may end up pairing empty vectors if all they contain are infinity pairs. Could that be an issue?

Comment (jayz22) [reply soroban-env-host/src/crypto/bn254.rs:326]: Rejecting empty inputs at the top level is just a specification choice (this also matches how evm specified it). Technically we could also let it be a no-op but if user passes in empty vectors it is likely an error so it makes sense to reject it. 
infinity points are still valid inputs for pairing (internal filtering is just an optimization). Pretty sure passing two infinity vectors should work without issue (it’s a good test case to add @sisuresh). We also meter them by  vector length treating them as regular points.

**Review by dmkozh — APPROVED** [review]
Please fix the typo.

**Review by jayz22 — APPROVED** [review]
Overall looks good, some minor cleanups and comments to address, but happy to merge it.
